### PR TITLE
8259 pic enabled

### PR DESCRIPTION
**Pull request by mrxx0**
We have a now a set up 8259 PIC.
The only entry we are handling is the keyboard.
To handle new hardware interrupt we need to add a new entry in the IDT and call the handling function for the interrupt.

### CODE REVIEW

**Comment by xlmod** [comment]
Other Interrupt don't work any more on this branch, do you known why?

File: kernel/idt.c [hunk 1]
```diff
@@ -6,15 +6,15 @@
  * Definition of ISR (Interrupt Service Routines) to handle exceptions
  * and interrupts
  *
- * created: 2022/10/18 - xlmod <glafond-@student.42.fr>
- * updated: 2022/10/19 - lfalkau <lfalkau@student.42.fr>
+ * created: 2022/10/20 - mrxx0 <chcoutur@student.42.fr>
+ * updated: 2022/10/20 - mrxx0 <chcoutur@student.42.fr>
```
Comment (xlmod) [inline kernel/idt.c:10]: Why changing the 'created' entry?

File: kernel/irq.c [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @arg(int_frame): interrupt frame structure.
+ */
+__attribute__ ((interrupt)) void keyboard_handler(t_int_frame *int_frame)
+{
+	kprintf("KEYBOARD IRQ\n");
+	LOAD_INTERRUPT_STACK;
+	pic_8259_eoi(1);
+	RESET_INTERRUPT_STACK;
+}
```
Comment (xlmod) [inline kernel/irq.c:29]: I think it's better to not doing anythink outside LOAD/RESET_INTERRUPT_STACK interval.

File: kernel/irq_internal.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+#ifndef IRQ_INTERNAL_H
+#define IRQ_INTERNAL_H
+
+#include "idt_internal.h"
+
+__attribute__ ((interrupt)) void keyboard_handler(t_int_frame *int_frame);
+
+#endif
```
Comment (xlmod) [inline kernel/irq_internal.h:19]: Like i said in previous comment i think it's preferable to regroup all prototype of interrupt handler in the same header file (idt_internal.h) like isr.

File: kernel/kernel.c [hunk 4]
```diff
@@ -22,6 +23,7 @@ extern struct vga vga;
  *
  */
 static void init_descriptor_tables() {
+	pic_8259_init(0x20, 0x28);
 	gdt_init();
 	idt_init();
```
Comment (xlmod) [inline kernel/kernel.c:28]: Preferable to use DEFINE set in pic header here.

File: kernel/pic_8259.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+/* Indicate the completion of the interrupt processing.
+ *
+ * @arg(irq_id) : the interrupt we want to indicate the end of processing
+ * */
+void pic_8259_eoi(uint8_t irq_id)
+{
+	if (irq_id > 8)
+		port_write(PIC2, PIC1);
+	port_write(PIC1, PIC1);
+}
```
Comment (xlmod) [inline kernel/pic_8259.c:83]: These writes are not understandable.
```c
#define PIC_EOI 0x20

...

if (irq_id)
   port_write(PIC2_COMMAND, PIC_EOI);
port_write(PIC1_COMMAND, PIC_EOI);
```
Even if the value of DEFINE are the same is better to use different one to understand what the function does.

**Review by xlmod — CHANGES_REQUESTED** [review]

**Comment by xlmod** [comment]
> 

to other interrupt not working or the different header file?

**Comment by xlmod** [comment]
> Isn't that answered in the PR message??

which question?

**Comment by mrxx0** [comment]
> Other Interrupt don't work any more on this branch, do you known why?

I will take a closer look at this issue.

**Comment by mrxx0** [comment]
> And i think is not a good idee to put irq handler prototype in a separate include file. Like isr i think is better to put handler on different source file but all the prototype in the same place in idt_internal.h as their are all handler to interrupt.

That's wthat I thought too, I will commit the changes to regroup the interrupt functions in the same file.

File: kernel/idt.c (hunk 1 above)
Comment (mrxx0) [reply kernel/idt.c:10]: We had an issue with the generated header.

File: kernel/irq.c (hunk 2 above)
Comment (mrxx0) [reply kernel/irq.c:29]: Ok but we must keep the EOI it's a requirement after an IRQ is fired by the PIC.

File: kernel/pic_8259.c (hunk 5 above)
Comment (mrxx0) [reply kernel/pic_8259.c:83]: It's the EOI (Enf Of Interrupt) this function tells PICs that the interrupt is processed and finished.
I will change the define here too with a more understandable ones.

**Comment by mrxx0** [comment]
> > Other Interrupt don't work any more on this branch, do you known why?
> 
> I will take a closer look at this issue.

Indeed we have a undesired behavior in the case 2 ISR are following.

Expected behavior : 
Divide Error then Page Fault should throw a Double Fault and halt the kernel

Actual behavior :
Divide Error then Page Fault don't throw a Double Fault and the kernel is still active.

Otherwise if we activate any ISR by calling them ourselves they are individually handled.

File: kernel/idt.c (hunk 1 above)
Comment (lfalkau) [reply kernel/idt.c:10]: My mistake, the [autoheader plugin](https://github.com/lfalkau/autoheader) was updating the 9th line (hardcoded), so if the file description was longer that one line, bad things occured.
This is now fixed, so make sure to pull latest changes for the plugin

File: kernel/irq.c (hunk 2 above)
Comment (lfalkau) [reply kernel/irq.c:29]: The frame we want to print is the caller frame, NOT the one we setup. So we should obviously print the frame BEFORE the `LOAD_INTERRUPT_STACK`.

In the same way, the `pic_8259_eoi` call will enable the PIC again to receive interrupts, so I think it's the last thing to do in our handler (after the `RESET_INTERRUPT_STACK`).

Those LOAD/RESET macros only resets stack segments registers, which are not used by regular functions (fix me if I'm wrong). The only thing that stores/restores the general purpose registers and thus must be called at the very top & at the very and are the `pusha`/`popa`/`iret` instructions, and this is handled well thanks to the interrupt attribute.

Comment (xlmod) [reply kernel/irq.c:29]: The Marco LOAD/RESET_INTERRUPT_STACK is not named correctly. In reality it's loading the kernel data segment. And the frame of the caller is push to the stack by the cpu (first argument of the function) so it's preferable to load the kernel data segment at first doing think to handle the interrupt and at the end reset the the previous segment.

**Review by xlmod — APPROVED** [review]
LGTM

**Pull request merged**